File: core/planner/src/main/java/ch/epfl/dias/calcite/adapter/pelago/RelPackingTraitDef.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to you under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ch.epfl.dias.calcite.adapter.pelago;

import org.apache.calcite.plan.RelOptPlanner;
import org.apache.calcite.plan.RelOptRuleCall;
import org.apache.calcite.plan.RelTraitDef;
import org.apache.calcite.plan.RelTraitSet;
import org.apache.calcite.rel.RelNode;
//import org.apache.calcite.rel.core.DeviceCross;
//import org.apache.calcite.rel.logical.LogicalDeviceCross;

/**
 * Definition of the device type trait.
 *
 * <p>Target device type is a physical property (i.e. a trait) because it can be
 * changed without loss of information. The converter to do this is the
 * {@link PelagoDeviceCross} operator.
 */
public class RelPackingTraitDef extends RelTraitDef<RelPacking> {
  public static final RelPackingTraitDef INSTANCE = new RelPackingTraitDef();

  protected RelPackingTraitDef() {}

  @Override public Class<RelPacking> getTraitClass() {
    return RelPacking.class;
  }

  @Override public String getSimpleName() {
    return "device";
  }

  @Override public RelNode convert(RelOptPlanner planner, RelNode rel, RelPacking to_packing,
                                   boolean allowInfiniteCostConverters) {
    if (rel.getConvention() != PelagoRel.CONVENTION()) return null;
    if (rel.getTraitSet().containsIfApplicable(to_packing)) return rel;

    RelNode p;
    if (to_packing == RelPacking.Packed) {
      p = PelagoPack  .create(rel, to_packing);
    } else {
      p = PelagoUnpack.create(rel, to_packing);
    }
    RelNode newRel = planner.register(p, rel);
    final RelTraitSet newTraitSet = rel.getTraitSet().replace(to_packing);
    if (!newRel.getTraitSet().equals(newTraitSet)) {
      newRel = planner.changeTraits(newRel, newTraitSet);
    }
    return newRel;
  }

  @Override public boolean canConvert(RelOptPlanner planner, RelPacking fromTrait, RelPacking toTrait) {
    return true;//fromTrait != toTrait;
  }

  @Override public RelPacking getDefault() {
    return RelPacking.UnPckd;
  }
}

// End RelDeviceTypeTraitDef.java

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/traits/RelPackingTraitDef.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.traits

import ch.epfl.dias.calcite.adapter.pelago.rel.{
  PelagoPack,
  PelagoRel,
  PelagoUnpack
}
import org.apache.calcite.plan.{RelOptPlanner, RelTraitDef}
import org.apache.calcite.rel.RelNode

object RelPackingTraitDef { val INSTANCE = new RelPackingTraitDef }

/**
  * Definition of the device type trait.
  *
  * Target device type is a physical property (i.e. a trait) because it can be
  * changed without loss of information. The converter to do this is the
  * [[PelagoPack]] and [[PelagoUnpack]] operator.
  */
class RelPackingTraitDef protected () extends RelTraitDef[RelPacking] {
  override def getTraitClass: Class[RelPacking] = classOf[RelPacking]

  override def getSimpleName = "device"

  override def convert(
      planner: RelOptPlanner,
      rel: RelNode,
      to_packing: RelPacking,
      allowInfiniteCostConverters: Boolean
  ): RelNode = {
    if (rel.getConvention ne PelagoRel.CONVENTION) return null
    if (rel.getTraitSet.containsIfApplicable(to_packing)) return rel
    var p: RelNode = null
    if (to_packing eq RelPacking.Packed) p = PelagoPack.create(rel, to_packing)
    else p = PelagoUnpack.create(rel, to_packing)
    var newRel = planner.register(p, rel)
    val newTraitSet = rel.getTraitSet.replace(to_packing)
    if (!(newRel.getTraitSet == newTraitSet))
      newRel = planner.changeTraits(newRel, newTraitSet)
    newRel
  }
  override def canConvert(
      planner: RelOptPlanner,
      fromTrait: RelPacking,
      toTrait: RelPacking
  ) = true //fromTrait != toTrait;
  override def getDefault: RelPacking = RelPacking.UnPckd
}
